app/domain/operations/hbx_enrollments/end_date_change.rb

```diff
@@ -0,0 +1,191 @@
+# frozen_string_literal: true
+
+require 'dry/monads'
+require 'dry/monads/do'
+module Operations
+  module HbxEnrollments
+    #This class updates the termination date of a HBX enrollment.
+    # If given termination date is less than enrollment termination -  it updates the enrollment
+    # If given termination date is greater than enrollment termination and is a prior year enrollment
+    # - it generates a new enrollment.
+    class EndDateChange
+      include Dry::Monads[:result, :do]
+
+      # @param [ EnrollmentId ] enrollment_id bson_id of a enrollment
+      # @param [ NewTerminationDate ] termination_date to update the enrollment
+      # @param [ EdiRequired ] edi_required to send the enrollment change to EDI
+      # @return [ Success ] :ok
+      def call(params:)
+        validated_params = yield validate(params)
+        @enrollment = yield fetch_enrollment(params["enrollment_id"])
+        _status = yield check_enrollment_eligibility(validated_params['new_termination_date'])
+        updated_enrollment = yield terminate_enrollment(validated_params)
+
+        end_date_after_effects(updated_enrollment, params["edi_required"].present?)
+
+        Success(:ok)
+      end
+
+      private
+
+      def validate(params)
+        return Failure("enrollment_id not present") if params["enrollment_id"].blank?
+        return Failure("new termination date not present") if params["new_termination_date"].blank?
+        Success(params)
+      end
+
+      def fetch_enrollment(enrollment_id)
+        Operations::HbxEnrollments::Find.new.call(id: enrollment_id)
+      end
+
+      def check_enrollment_eligibility(termination_date)
+        @termination_date = Date.strptime(termination_date, "%m/%d/%Y")
+        eligible_states = %w[coverage_terminated coverage_termination_pending]
+        eligible_states << 'coverage_expired' if EnrollRegistry[:change_end_date].settings(:expired_enrollments).item
+
+        return Failure('Enrollment not in valid state') unless eligible_states.include?(@enrollment.aasm_state)
+
+        return Failure('Invalid termination date') if termination_date_check_fails?
+
+        return Failure("Overlapping coverage exists") if check_if_overlapping_coverage_exists?
+        Success(:ok)
+      end
+
+      def termination_date_check_fails?
+        prior_or_current_py_enr = @enrollment.prior_plan_year_coverage? || @enrollment.active_plan_year_coverage?
+
+        if prior_or_current_py_enr
+          term_or_expiration_date = @enrollment.is_shop? ? @enrollment.sponsored_benefit_package.end_on : @enrollment.effective_on.end_of_year
```

Comment:
can you handle enrollment when you don't have sponsored_benefit_package on it

Reply:
I can add a check, but we will have having only active, termed or expired enrollments shown in this case, so I think there will not be any enrollment which will not have sponsored benefit package in this case